src/Options.php

```diff
@@ -0,0 +1,307 @@
+
+        return $this;
+    }
+
+    /**
+     * Disable DNS pinning.
+     *
+     * @return Options
+     */
+    public function disablePinDns()
+    {
+        $this->pinDns = false;
+
+        return $this;
+    }
+
+    /**
+     * Checks if a specific value is in a list.
+     *
+     * @param 'whitelist'|'blacklist' $listName
+     * @param string $type
+     * @param $value
+     *
+     * @throws InvalidOptionException
+     *
+     * @return bool
+     */
+    public function isInList($listName, $type, $value)
+    {
+        $this->checkListByName($listName);
+
+        if (!array_key_exists($type, $this->lists[$listName])) {
+            throw new InvalidOptionException('Provided type "'.$type.'" must be "ip", "port", "domain" or "scheme"');
+        }
+
+        if (empty($this->lists[$listName][$type])) {
+            return 'whitelist' === $listName;
+        }
+
+        //For domains, a regex match is needed
+        if ('domain' === $type) {
+            foreach ($this->lists[$listName][$type] as $domain) {
+                if (preg_match('/^'.$domain.'$/i', $value)) {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        return in_array($value, $this->lists[$listName][$type], true);
+    }
+
+    /**
+     * Returns a specific list.
+     *
+     * @param 'whitelist'|'blacklist' $listName
+     * @param string|null $type
+     *
+     * @throws InvalidOptionException
+     *
+     * @return array
+     */
+    public function getList($listName, $type = null)
+    {
+        $this->checkListByName($listName);
+
+        if (null !== $type) {
+            if (!array_key_exists($type, $this->lists[$listName])) {
+                throw new InvalidOptionException('Provided type "'.$type.'" must be "ip", "port", "domain" or "scheme"');
+            }
+
+            return $this->lists[$listName][$type];
+        }
+
+        return $this->lists[$listName];
+    }
+
+    /**
+     * Sets a list, the values must be passed as an array.
+     *
+     * @param 'whitelist'|'blacklist' $listName
+     * @param array $values
+     * @param string|null $type
+     *
+     * @throws InvalidOptionException
+     *
+     * @return Options
+     */
+    public function setList($listName, $values, $type = null)
+    {
+        $this->checkListByName($listName);
+
+        if (!is_array($values)) {
+            throw new InvalidOptionException('Provided values must be an array, "'.gettype($values).'" given');
+        }
+
+        if (null !== $type) {
+            if (!array_key_exists($type, $this->lists[$listName])) {
+                throw new InvalidOptionException('Provided type "'.$type.'" must be "ip", "port", "domain" or "scheme"');
+            }
+
+            $this->lists[$listName][$type] = $values;
+
+            return $this;
+        }
+
+        foreach ($values as $type => $value) {
+            if (!in_array($type, ['ip', 'port', 'domain', 'scheme'], true)) {
```

Comment:
Can't this list of type be define as a constant so you can re-use it in exception message.
And if in case we are adding a new type we won't have to update each exception message but the constant.

Reply:
I can't use a constant for an array, I use a static variable, and I removed duplication on exception message by using static methods on the InvalidOptionException class, WDYT ?

Reply:
Oh yeah I forgot about the constant & array, it's ok like this 👍